### DISCUSSION TITLE
Question: observable array push does not react.

### DISCUSSION THREAD

**Discussion by HwangTaehyun** [mobxjs/mobx#2738] [Q&A] [answered]
Hi, I use mobx everyday!

Recently, I don't understand following autorun's working logic. Please help me..

```js
import { autorun, observable, makeAutoObservable } from 'mobx'

class Message {
    title
    author
    likes
    constructor(title, author, likes) {
        makeAutoObservable(this, {likes: observable.shallow})
        this.title = title
        this.author = author
        this.likes = likes
    }

    updateTitle(title) {
        this.title = title
    }

    pushLikes(item) {
        this.likes.push(item);
    }
}

let message = new Message("Foo", { name: "Michel" }, ["Joe", "Sara"])

autorun(() => {
    console.log(message.title)
})

autorun(() => {
    // case (1) is called
    console.log(message.likes.slice())
    // case (2) is called
    message.likes.map((item)=>{
        console.log(item);
    })
    // case (3) is not called
    console.log(message.likes);
})

message.pushLikes("Taehyun");
```

I expect that when I call Message class's "pushLikes" method then (3) should be called because Message class's likes array is observed by shallow strategy.

I understand shallow observe means that tracking whether observable array's each item's reference has changed or new item is added.

Why does not case (3) react?

Maybe this has sth to do with below. ([https://mobx.js.org/understanding-reactivity.html](https://mobx.js.org/understanding-reactivity.html))
![image](https://user-images.githubusercontent.com/42465090/105443712-ba538500-5caf-11eb-81c0-561b207fb6f6.png)

Changing the question, how should I use the observe step of an observable array in what situations?

Thank you for reading my question!!

**Comment by danielkcz** [comment]
Please, can you provide runnable reproduction? It's kinda to understand what's wrong.

**Reply by HwangTaehyun** [reply]
Sorry. It seems that the comment was wrong in the middle and it could not be executed. Now the code is running.

**Comment by mweststrate** [comment, marked as answer]
See
https://mobx.js.org/understanding-reactivity.html#common-pitfall-consolelog

**Reply by HwangTaehyun** [reply]
My question is, if the observable array of the mobx store is shallow observable, I think the following action function should react when i push data to the observable array.
```
autorun(() => {
    console.log(message.likes)
})
```
The common pitfall docs said this wasn't working, but I'm curious why it didn't work for a shallow observable array.

![image](https://user-images.githubusercontent.com/42465090/106221815-a82e9500-6221-11eb-93e9-4c6ac3409536.png)

The meaning of shallow observe is to see whether the reference of items has changed with respect to the first depth, and when the item is pushed, there is a change in the first depth, so I thought that a reaction should occur of course.

Thank you.

**Reply by mweststrate** [reply]
the only observable used in that autorun is the "likes" property of "messages". That isn't modified as it point to the same thing, so no reaction. If you would be using the "length" property of "likes", and you add an item, that does cause the "length" to change, so you get a reaction. Etc. Autorun reacts to the observables you actually _read_

**Reply by HwangTaehyun** [reply]
Thank you for your explanation. There seems to be something I was wrong about shallow observable. I need to study in more detail. Thank you.
